### PR TITLE
Documenting plotting and sensitivity functions

### PR DESCRIPTION
**Pull request by svavil**
I have completed the vignettes for visualization and sensitivity analysis functions. As soon as we finalize these changes, a separate PR will follow to re-build the HTML documentation files from the RMD vignettes.

### CODE REVIEW

File: vignettes/example-visualization.Rmd [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+```{r datacombined}
+dataCombined <- DataCombined$new()
+dataCombined$addDataSets(observedData, names = "Observed", groups = "Aciclovir")
+dataCombined$addSimulationResults(simulatedScenarios$TestScenario$results, names = "Simulated", groups = "Aciclovir")
+dataCombined$toDataFrame()
+```
+
+The simulation results are stored in a list returned by the `runScenarios()` function. Plotting and visualization is performed by storing these results along with matching observed data in an instance of the `DataCombined` class and passing it to plotting functions.
+
+Plotting functions in the `{ospsuite}` package are wrappers around `{tlf}` plotting functions that provide default plot configuration options. All of them accept instances of `DataCombined` class as the data source.
```
Comment (IndrajeetPatil) [inline vignettes/example-visualization.Rmd:52]: I am not sure if we need to mention anything about implementation details here; i.e. about `{tlf}`.

File: vignettes/example-visualization.Rmd [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+The resulting plots (returned by plotting functions) are instances of `ggplot` class. They can be modified and saved to files with `{ggplot2}` functions:
+
+```{r save plots}
+plotObject <- plotIndividualTimeProfile(dataCombined)
+ggplot2::ggsave(filename = "../Results/aciclovir_time_profile.png", plotObject, width = 8, height = 4)
+```
+
+### Troubleshooting
+
+* At any time, you can check the groups assigned to the datasets in the `DataCombined` class instance by examining the output of `dataCombined$toDataFrame()`.
```
Comment (IndrajeetPatil) [inline vignettes/example-visualization.Rmd:85]:
```suggestion
* At any time, you can check the groups assigned to the datasets in the `DataCombined` object by examining the output of `dataCombined$toDataFrame()`.
```

File: vignettes/example-visualization.Rmd [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+`DataCombined` is a class used to store matching observed and simulated data. Initialize a new class instance and populate it with data with the following code:
+
+```{r datacombined}
+dataCombined <- DataCombined$new()
+dataCombined$addDataSets(observedData, names = "Observed", groups = "Aciclovir")
+dataCombined$addSimulationResults(simulatedScenarios$TestScenario$results, names = "Simulated", groups = "Aciclovir")
+dataCombined$toDataFrame()
+```
+
+The simulation results are stored in a list returned by the `runScenarios()` function. Plotting and visualization is performed by storing these results along with matching observed data in an instance of the `DataCombined` class and passing it to plotting functions.
```
Comment (IndrajeetPatil) [inline vignettes/example-visualization.Rmd:50]:
```suggestion
The simulation results are stored in a list returned by the `runScenarios()` function. Plotting and visualization is performed by storing these results along with matching observed data in a `DataCombined` object and passing it to plotting functions.
```

File: vignettes/example-workflow.Rmd [hunk 4]
```diff
@@ -11,20 +11,35 @@ vignette: >
...
+* the `ProjectConfiguration.xlsx` stores paths to four configuration files: `Parameters/IndividualPhysiology.xlsx`, `Parameters/ApplicationParameters.xlsx`, `Parameters/Scenarios.xlsx` and `Parameters/IndividualParameters.xlsx`
+* the `IndividualPhysiology.xlsx` file defines an individual called `Indiv1`: an aged 30, 73 kg and 176 cm Caucasian human male.
+* the `ApplicationParameters.xlsx` defines a protocol called `Aciclovir_iv_250mg`: a single dose of 250 mg delivered to the `Applications|IV 250mg 10min|Application_1|ProtocolSchemaItem` container.
+* the `Scenarios.xlsx` defines a scenario called `TestScenario`, using `Indiv1` as an individual and `Aciclovir_iv_250mg` as the application protocol. The model file `Aciclovir.pkml` is also referenced and is expected to exist in the `projectConfiguration$modelFolder` folder. 
+* the `IndividualParameters.xlsx` file redefines the GFR to be 90 mL/min
 
-```{r}
+Scenarios are then executed by creating a `ScenarioConfiguration` class instance and calling a `runScenarios()` function with a specific scenario name (`TestScenario` in this case):
```
Comment (IndrajeetPatil) [inline vignettes/example-workflow.Rmd:40]:
```suggestion
Scenarios are then executed by creating a `ScenarioConfiguration` object and calling a `runScenarios()` function with a specific scenario name (`TestScenario` in this case):
```

File: vignettes/example-workflow.Rmd [hunk 5]
```diff
@@ -40,6 +55,18 @@ The `runScenarios()` function relies on an `OutputPaths` variable being defined
 
 The `runScenarios()` function returns a named list of lists with `Simulation` class instances, `SimulationResults` class instances, and vectors of output values.
```
Comment (IndrajeetPatil) [inline vignettes/example-workflow.Rmd:56]:
```suggestion
The `runScenarios()` function returns a named list of lists with `Simulation` objects, `SimulationResults` objects, and vectors of output values.
```

File: vignettes/example-workflow.Rmd [hunk 6]
```diff
@@ -11,20 +11,35 @@ vignette: >
...
-  comment = "#>",
-  eval = FALSE
+  comment = "#>"
+)
+knitr::opts_knit$set(
+  root.dir = "../tests/data/TestProject/Code/"
 )
 ```
 
 ```{r, echo = FALSE, results = "hide", message = F}
```
Comment (IndrajeetPatil) [inline vignettes/example-workflow.Rmd:21]: Same comments as above.

File: vignettes/sensitivity-analysis.Rmd [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+  scenarioConfiguration = scenarioConfiguration,
+  customParams = NULL, saveSimulationsToPKML = FALSE
+)
+```
+
+### Workflow
+
+Sensitivity analysis quantifies how the pharmacokinetics of the drug changes with a variation in simulation parameters. This is important to track if the values of simulation parameters are uncertain.
+
+In the aciclovir simulation example, the lipophilicity of aciclovir was assumed to be -0.100 in log units. In the sensitivity analysis, we want to quantify how much the pharmacokinetic parameters change if the lipophilicity of aciclovir varies in the (-0.01 .. -1.00) range.
```
Comment (IndrajeetPatil) [inline vignettes/sensitivity-analysis.Rmd:41]:
```suggestion
In the aciclovir simulation example, the lipophilicity of aciclovir was assumed to be -0.100 in log units. In the sensitivity analysis, we want to quantify how much the pharmacokinetic parameters change if the lipophilicity of aciclovir varies in the (-0.01, -1.00) range.
```

File: vignettes/sensitivity-analysis.Rmd [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+
+The `sensitivityCalculation()` function in the `{esqlabsR}` package does that by re-running the simulation with a set of updated parameter values. The function returns a list with output paths, parameter paths, an instance of `SimulationResults` class and a data frame with computed pharmacokinetic parameters for each of the input parameter values.
```
Comment (IndrajeetPatil) [inline vignettes/sensitivity-analysis.Rmd:43]:
```suggestion
The `sensitivityCalculation()` function in the `{esqlabsR}` package does that by re-running the simulation with a set of updated parameter values. The function returns a list with output paths, parameter paths, a `SimulationResults` object, and a data frame with computed pharmacokinetic parameters for each of the input parameter values.
```

File: vignettes/sensitivity-analysis.Rmd [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+```{r analysis}
+simulation <- simulatedScenarios$TestScenario$simulation
+OutputPaths <- enum(list(
+  Aciclovir_PVB = "Organism|PeripheralVenousBlood|Aciclovir|Plasma (Peripheral Venous Blood)"
+))
+analysis <- sensitivityCalculation(simulation, OutputPaths, parameterPaths = "Aciclovir|Lipophilicity")
+head(analysis$pkData)
+```
+
+In the aciclovir example case, the default value of lipophilicity was -0.100 log units, corresponding to the area under curve of 3915.87 µmol×min/L. Changing the lipophilicity to -0.01 log units leads to a decrease of AUC to 3895.43 µmol×min/L (a change of -0.52%), while changing the lipophilicity to -1.00 log units leads to an increase of AUC to 4015.73 µmol×min/L (a change of 2.55%).
```
Comment (IndrajeetPatil) [inline vignettes/sensitivity-analysis.Rmd:54]:
```suggestion
In the aciclovir example case, the default value of lipophilicity was -0.100 log units, corresponding to the area under curve (AUC) of 3915.87 µmol×min/L. Changing the lipophilicity to -0.01 log units leads to a decrease of AUC to 3895.43 µmol×min/L (a change of -0.52%), while changing the lipophilicity to -1.00 log units leads to an increase of AUC to 4015.73 µmol×min/L (a change of 2.55%).
```

File: vignettes/sensitivity-analysis.Rmd [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+The results of the sensitivity analysis can be plotted with the `sensitivitySpiderPlot()` function.
+
+```{r plot, fig.width=7, fig.height=4}
+sensitivitySpiderPlot(analysis)
```
Comment (IndrajeetPatil) [inline vignettes/sensitivity-analysis.Rmd:59]: Why isn't `sensitivityTimeProfiles()` mentioned?

https://esqlabs.github.io/esqlabsR/reference/sensitivityTimeProfiles.html

File: vignettes/sensitivity-analysis.Rmd [hunk 11]
```diff
@@ -0,0 +1,70 @@
+---
+title: "Sensitivity analysis"
+output: rmarkdown::html_vignette
+#output: pdf_document
```
Comment (IndrajeetPatil) [inline vignettes/sensitivity-analysis.Rmd:4]:
```suggestion
```

Comment (IndrajeetPatil) [reply vignettes/sensitivity-analysis.Rmd:4]: Not necessary.

File: vignettes/example-visualization.Rmd [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+plotResidualsVsSimulated(dataCombined)
+```
+
+```{r residuals vs time, fig.width=7, fig.height=4}
+plotResidualsVsTime(dataCombined)
+```
+
+The resulting plots (returned by plotting functions) are instances of `ggplot` class. They can be modified and saved to files with `{ggplot2}` functions:
+
+```{r save plots}
```
Comment (IndrajeetPatil) [inline vignettes/example-visualization.Rmd:78]:
```suggestion
```{r save-plots}
```

cf. https://github.com/Open-Systems-Pharmacology/Suite/blob/develop/CODING_STANDARDS_R.md#rmarkdown

Comment (IndrajeetPatil) [reply vignettes/example-visualization.Rmd:78]: Please change all chunk labels accordingly.

**Review by IndrajeetPatil — CHANGES_REQUESTED** [review]
Looks great to me! Thanks, @svavil! 

I just had a few minor corrections here and there, and a few comments.

File: vignettes/example-visualization.Rmd [hunk 13]
```diff
@@ -0,0 +1,96 @@
...
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+knitr::opts_knit$set(
+  root.dir = "../tests/data/TestProject/Code/"
+)
+```
+
+```{r, echo = FALSE, results = "hide", message = F}
```
[the remark this answers is not in the document]
Comment (svavil) [reply vignettes/example-visualization.Rmd:21]: Some code blocks are in `echo = FALSE`. The idea is that these objects have been created in previous vignettes, so they line up in a sequence of vignettes actually.

File: vignettes/sensitivity-analysis.Rmd (hunk 10 above)
Comment (svavil) [reply vignettes/sensitivity-analysis.Rmd:59]: I didn't make `sensitivityTimeProfiles` work properly. Is there a working example that you have somewhere?

File: vignettes/example-visualization.Rmd (hunk 1 above)
Comment (svavil) [reply vignettes/example-visualization.Rmd:52]: I think it doesn't hurt, even if mentioning `{tlf}` is not directly beneficial to the users.

File: vignettes/sensitivity-analysis.Rmd (hunk 10 above)
Comment (IndrajeetPatil) [reply vignettes/sensitivity-analysis.Rmd:58]: https://esqlabs.github.io/esqlabsR/reference/sensitivityTimeProfiles.html#ref-examples

File: vignettes/example-visualization.Rmd [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+dataCombined$addDataSets(observedData, names = "Observed", groups = "Aciclovir")
+dataCombined$addSimulationResults(simulatedScenarios$TestScenario$results, names = "Simulated", groups = "Aciclovir")
+dataCombined$toDataFrame()
+```
+
+The simulation results are stored in a list returned by the `runScenarios()` function. Plotting and visualization is performed by storing these results along with matching observed data in a `DataCombined` object and passing it to plotting functions.
+
+Plotting functions in the `{ospsuite}` package are wrappers around `{tlf}` plotting functions that provide default plot configuration options. All of them accept instances of `DataCombined` class as the data source. 
+
+Time profile plots visualize the pharmacokinetics of the drug in question and help assess if the observed data (shown with scatter plots) match the simulated data (shown with line plots).
```
Comment (IndrajeetPatil) [inline vignettes/example-visualization.Rmd:54]:
```suggestion
Time profile plots visualize the pharmacokinetics of the drug in question and help assess if the observed data (represented by points and error bars) match the simulated data (represented by lines).

```

File: vignettes/example-visualization.Rmd [hunk 15]
```diff
@@ -0,0 +1,96 @@
...
+
+```{r plot time profile, fig.width=7, fig.height=4}
+plotIndividualTimeProfile(dataCombined)
+```
+
+Observed vs simulated plots show if simulated time points and observed time points follow a linear trend.
```
Comment (IndrajeetPatil) [inline vignettes/example-visualization.Rmd:60]:
```suggestion
Observed versus simulated plots show if simulated time points and observed time points follow a linear trend.
```

File: vignettes/example-visualization.Rmd [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+
+```{r residuals vs simulated, fig.width=7, fig.height=4}
+plotResidualsVsSimulated(dataCombined)
+```
+
+```{r residuals vs time, fig.width=7, fig.height=4}
+plotResidualsVsTime(dataCombined)
+```
+
+The resulting plots (returned by plotting functions) are instances of `ggplot` class. They can be modified and saved to files with `{ggplot2}` functions:
```
Comment (IndrajeetPatil) [inline vignettes/example-visualization.Rmd:76]:
```suggestion
The plots returned by plotting functions are `ggplot` objects. They can be modified further and saved to files with `{ggplot2}` functions:
```

File: vignettes/example-visualization.Rmd (hunk 12 above)
Comment (IndrajeetPatil) [reply vignettes/example-visualization.Rmd]: This should be resolved only when all chunk labels have been renamed to follow our coding conventions.

File: vignettes/sensitivity-analysis.Rmd [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+knitr::opts_chunk$set(
+  collapse = TRUE,
+  comment = "#>"
+)
+knitr::opts_knit$set(
+  root.dir = "../tests/data/TestProject/Code/"
+)
+```
+
+```{r, echo = FALSE, results = "hide", message = F}
```
Comment (IndrajeetPatil) [inline vignettes/sensitivity-analysis.Rmd:20]:
```suggestion
```{r, echo = FALSE, results = "hide", message = FALSE}
```

**Pull request merged**